crypto/multisig/src/lib.rs

```diff
@@ -0,0 +1,245 @@
+use prost::Message;
+use serde::{Deserialize, Serialize};
+
+/// The maximum number of signatures that can be included in a multi-signature.
+pub const MAX_SIGNATURES: usize = 10;
+
+/// A multi-signature: a collection of one or more signatures.
+#[derive(Clone, Deserialize, Digestible, Eq, Message, PartialEq, Serialize)]
+pub struct MultiSig<
+    S: Clone + Default + Digestible + Eq + Message + PartialEq + Serialize + Signature,
+> {
+    #[prost(message, repeated, tag = "1")]
+    signatures: Vec<S>,
+}
+
+impl<S: Clone + Default + Digestible + Eq + Message + PartialEq + Serialize + Signature>
+    MultiSig<S>
+{
+    /// Construct a new multi-signature from a collection of signatures.
+    pub fn new(signatures: Vec<S>) -> Self {
+        Self { signatures }
+    }
+}
+
+/// A set of M-out-of-N public keys.
+#[derive(Clone, Deserialize, Digestible, Eq, Message, PartialEq, Serialize)]
+#[serde(bound = "")]
+pub struct SignerSet<P: Default + PublicKey + Message> {
+    /// List of potential signers.
+    #[prost(message, repeated, tag = "1")]
+    signers: Vec<P>,
+
+    /// Minimum number of signers required.
+    #[prost(uint32, tag = "2")]
+    threshold: u32,
+}
+
+impl<P: Default + PublicKey + Message> SignerSet<P> {
+    /// Construct a new `SignerSet` from a list of public keys and threshold.
+    pub fn new(signers: Vec<P>, threshold: u32) -> Self {
+        Self { signers, threshold }
+    }
+
+    /// Verify a message against a multi-signature, returning the list of
+    /// signers that signed it.
+    pub fn verify<
+        S: Clone + Default + Digestible + Eq + Message + PartialEq + Serialize + Signature,
+    >(
+        &self,
+        message: &[u8],
+        multi_sig: &MultiSig<S>,
+    ) -> Result<Vec<P>, SignatureError>
+    where
+        P: Verifier<S>,
+    {
+        // If the signature contains less than the threshold number of signers or more
+        // than the hardcoded limit, there's no point in trying.
+        if multi_sig.signatures.len() < self.threshold as usize
+            || multi_sig.signatures.len() > MAX_SIGNATURES
+        {
+            return Err(SignatureError::new());
+        }
+
+        let mut matched_signers = Vec::new();
+        let mut potential_signers = self.signers.clone();
+        for signature in multi_sig.signatures.iter() {
+            let matched_signer = potential_signers.iter().find_map(|signer| {
+                signer
+                    .verify(message, signature)
+                    .ok()
+                    .map(|_| signer.clone())
+            });
+            if let Some(matched_signer) = matched_signer {
+                potential_signers.retain(|signer| signer != &matched_signer);
+                matched_signers.push(matched_signer);
+            }
+        }
+
+        if matched_signers.len() < self.threshold as usize {
+            return Err(SignatureError::new());
+        }
+
+        Ok(matched_signers)
+    }
+}
+
+#[cfg(test)]
+mod test {
```

Comment:
I'd go a bit heavier on unit testing situations where the signer set and/or the signature contains duplicates. It looks like there are a few places where we are using Vec instead of HashSet because of no_std, so we are more prone to bugs related to duplicates than usual.